// tt-db/src/BaseDao.js
/**
 * Desc: dao基类。
 * @author zheng.xiaojun
 * @constructor
 */
function BaseDao(){
    this.Po = null;
    this._func2Dao = function(client, func, args){
        var arr = Array.apply(null, args);
        arr.splice(0, 1, this.Po);
        func.apply(client, arr);
    }
    this.insert = function(client, args, cb){
        this._func2Dao(client, client.insert, arguments);
    };
    this.update = function(client, values, cnd, args, cb){
        this._func2Dao(client, client.update, arguments);
    };
    this.select = function(client, cnd, args, cb){
        this._func2Dao(client, client.select, arguments);
    };
    this.del = function(client, cnd, args, cb){
        this._func2Dao(client, client.del, arguments);
    };
    this.list = function(client, cnd, args, cb){
        this._func2Dao(client, client.list, arguments);
    };
    this.listPage = function(client, page, num, cnd, args, cb){
        this._func2Dao(client, client.listPage, arguments);
    };
}

module.exports = BaseDao;
