// tt-db/src/dbPool.js
var _poolModule = require('generic-pool');
var queues = require('mysql-queues');
const DEBUG = true;
var escape = require("mysql").escape;
var escapeId = require("mysql").escapeId;


function queryFormat (sql, values, stringifyObjects, timeZone) {
    if (!values) return sql;
    if(values instanceof Array){
        values = [].concat(values);
        return sql.replace(/\?\??/g, function(match) {
            if (!values.length) return match;
            if (match == "??") return escapeId(values.shift());
            return escape(values.shift(), stringifyObjects, timeZone);
        });
    }

    return sql.replace(/\:(\w+)/g, function (txt, key) {
        if (values.hasOwnProperty(key)) return this.escape(values[key]);
        return txt;
    }.bind(this));
}
/*
 * Create mysql connection pool.
 */
var create = function(cnnCfg) {
	return _poolModule.Pool({
		name: cnnCfg.name || 'mysql',
		create: function(callback) {
            var mysql = require(cnnCfg.dbModule || 'mysql');
			var client = mysql.createConnection({
				host: cnnCfg.host,
				user: cnnCfg.user,
				password: cnnCfg.password,
				database: cnnCfg.database,
                queryFormat : queryFormat,
                debug : cnnCfg.debug || false
			});
            queues(client, DEBUG);
			callback(null, client);
		},
		destroy: function(client) {
			client.destroy();
		},
		max: 10,
		idleTimeoutMillis : 30000,
		log : false
	});
};

exports.create = create;
